packages/client/src/rpc.ts

```diff
@@ -0,0 +1,155 @@
+  Transaction, // A normal transaction with data modification
+  TransientTransaction // A transient transaction with derived data modification.
+}
+
+interface PromiseInfo {
+  resolve: (value?: any) => void
+  reject: (error: any) => void
+}
+export interface RawClient {
+  request: <R>(method: string, ...params: any[]) => Promise<R>
+  addEventListener: (type: EventType, listener: EventListener) => void
+  close: () => void
+}
+
+export function newRawClient (token: string, host: string, port: number): RawClient {
+  const requests = new Map<ReqId, PromiseInfo>()
+  let lastId = 0
+
+  function createWebsocket (): Promise<WebSocket> { // eslint-disable-line @typescript-eslint/promise-function-async
+    return new Promise<WebSocket>((resolve) => {
+      // Let's sure token is valid one
+      const WebSocket = (typeof window !== 'undefined' ? window.WebSocket : require('ws'))
+      const ws = new WebSocket(`ws://${host}:${port}/${token}`)
+
+      ws.onopen = () => {
+        resolve(ws)
+      }
+
+      ws.onerror = (ev: Event) => {
+        console.log('websocket error: ', ev)
+        // reject all pending requests
+        for (const info of requests.values()) {
+          info.reject(ev)
+        }
+        requests.clear()
+      }
+
+      ws.onmessage = (ev: MessageEvent) => {
+        const response = readResponse(ev.data)
+        if (response.id === undefined) {
+          if (response.result !== undefined) {
+            for (const listener of (listeners.get(EventType.Transaction) ?? [])) {
+              listener(response.result)
+            }
+          }
+        } else {
+          const promise = requests.get(response.id)
+          if (promise !== undefined) {
+            if (response.error !== undefined) {
+              promise.reject(response.error)
+            } else {
+              promise.resolve(response.result)
+            }
+            requests.delete(response.id)
+          } else {
+            throw new Error('unknown rpc id')
+          }
+        }
+        if ((response.clientTx !== undefined) && response.clientTx.length > 0) {
+          for (const listener of (listeners.get(EventType.TransientTransaction) ?? [])) {
+            listener(response.clientTx)
+          }
+        }
+      }
+    })
+  }
+
+  let websocket: WebSocket | undefined
+
+  async function getWebSocket (): Promise<WebSocket> {
+    const WebSocket = (typeof window !== 'undefined' ? window.WebSocket : require('ws'))
+    if (websocket === undefined || (websocket.readyState === WebSocket.CLOSED || websocket.readyState === WebSocket.CLOSING)) {
+      websocket = await createWebsocket()
+    }
+    return websocket
+  }
+
+  async function request<R> (method: string, ...params: any[]): Promise<R> {
+    return await new Promise<any>((resolve, reject) => {
+      const id = ++lastId
+      requests.set(id, {
+        resolve,
+        reject
+      })
+        getWebSocket().then((ws: WebSocket) => { // eslint-disable-line
```

Comment:
Formatting seems to be invalid due to `eslint-disable-line`

Reply:
Fixed.